python/pyarrow/serialization.py

```diff
@@ -67,9 +68,11 @@ def _deserialize_default_dict(data):
 
     serialization_context.register_type(
         type(lambda: 0), "function",
-        pickle=True)
+        custom_serializer=pickle.dumps, custom_deserializer=pickle.loads)
 
-    serialization_context.register_type(type, "type", pickle=True)
+    serialization_context.register_type(type, "type",
+                                        custom_serializer=pickle.dumps,
+                                        custom_deserializer=pickle.loads)
```

Comment:
If no custom serializer/deserializer is passed, should this default to `pickle`?

Reply:
Currently it defaults to using __dict__, which is more efficient so it seems a good default, we use that in Ray for pretty much all types (however rarely I have seen cases where it doesn't work). I'd prefer to keep __dict__ the default but no big deals since it can easily be changed.

Reply:
What about lambdas?

Reply:
In Ray, lambdas are handled with cloudpickle as are builtin types like "type"; most user defined types are handled via __dict__; namedtuples are special cased.

One reasonable way to handle this is to provide default callbacks for types we encounter that don't work with __dict__ (using cloudpickle or a custom serializer/deserializer), there are already examples of this in the repo, and use __dict__ for anything else. That worked well for us so far. If this strategy doesn't work for the user, they have full flexibility by providing their own serialization context.

What do you think?

Reply:
This is sufficiently advanced stuff that I think it's not unreasonable to ask people to explicitly use cloudpickle in a case like this. I would be OK if we changed this patch to do that to get the build passing, and we can always move on later